[air-doc] Rework experiment tracking docs for Torch trainer. (#38684)
ray-project/ray: doc/source/train/user-guides/experiment-tracking.rst

Author: yeah got it. I have the same feeling. @angelinalg maybe we can sit down some time to work on this?

.. _train-experiment-tracking-native:

==================================
Experiment Tracking with Ray Train
==================================

.. note::
    This guide is relevant for all DataParallelTrainer, including TorchTrainer and TensorflowTrainer.

Most experiment tracking libraries work out-of-the-box with Ray Train. 
In the following guide, you will learn how to use your favorite experiment tracking libraries 
while doing distributed data parallel training with Ray Train. 

Ray Train lets you use native experiment tracking libraries by customizing your tracking 
logic inside ``train_loop_per_worker``. 
This gives you the highest degree of freedom to decide when and where to log params, metrics, 
models or even media contents. 
This also allows you to use some native integrations that these tracking frameworks have with 
specific training frameworks, for example ``mlflow.pytorch.autolog()``, 
``lightning.pytorch.loggers.MLFlowLogger`` etc. 
In this way, you get to port your experiment tracking logic to Ray Train with minimal change. 

How to set up your code to log distributed training experiment
==============================================================

> Step 1: Decide a tracking service to use

Author: this is actually about tracking service than shared storage. so I am gonna leave it as is.

----------------------------------------

> For easy configuration, we recommend the following:

Reviewer: Use bullet points in this section?

MLflow: 

> file based tracking uri (``mlflow.start_run(tracking_uri=”file:some_shared_storage_location”)``)

Author: done!

> file based tracking uri (``mlflow.start_run(tracking_uri=”file:some_shared_storage_location”)``)

Author: yes, two options: online vs offline modes.

> file based tracking uri (``mlflow.start_run(tracking_uri=”file:some_shared_storage_location”)``)

Author: option of using databricks is actually not about shared storage anymore. shared storage is about offline mode online, whereas using databricks is online mode.
I reworded a session a little bit. Can you see if it's clearer to you?

> or externally hosted by Databricks(``mlflow.start_run(tracking_uri=”databricks”)``).

Author: done!

TensorBoard: 

A shared storage location where all nodes can write to.

> W&B:

Author: I think this is ok and people can understand it fine.

> Should work out-of-the-box.

Reviewer: Shall we also mention wandb online/offline mode here? Online mode will push to the serving endpoint, and offline mode saves the artifact locally and users can upload it afterwards with `wandb sync`.

> Should work out-of-the-box.

Author: like I mentioned in the other comment, externally hosted services are considered online mode and shared storage is irrelevant.

> Step 2: Make sure that all nodes and processes have access to credentials

Author: Done!

-------------------------------------------------------------------------

See example below.

> Step3: Initialize the run 

Author: Done

-------------------------

Ray Train provides a training context from where you can grab the 
training id (:meth:`context.get_trial_id() <ray.train.context.TrainContext.get_trial_id>`) 
and name (:meth:`context.get_trial_name() <ray.train.context.TrainContext.get_trial_name>`) 
if desired. 

> Step4: Log

Author: if we are anchoring on "flow" rather than "difference", we should still keep this as "Log"?

.. _train-experiment-tracking-native:

==================================
Experiment Tracking with Ray Train
==================================

.. note::
    This guide is relevant for all DataParallelTrainer, including TorchTrainer and TensorflowTrainer.

Most experiment tracking libraries work out-of-the-box with Ray Train. 
In the following guide, you will learn how to use your favorite experiment tracking libraries 
while doing distributed data parallel training with Ray Train. 

The guide will be focusing on TorchTrainer although the ideas should apply to DataParallelTrainer
in general.

Ray Train lets you use native experiment tracking libraries by customizing your tracking 
logic inside ``train_loop_per_worker``. 
This gives you the highest degree of freedom to decide when and where to log params, metrics, 
models or even media contents. 
This also allows you to use some native integrations that these tracking frameworks have with 
specific training frameworks, for example ``mlflow.pytorch.autolog()``, 
``lightning.pytorch.loggers.MLFlowLogger`` etc. 

> In this way, you get to port your experiment tracking logic to Ray Train with minimal change. 

Author: I actually think this is the essence to grab. Basically put whatever logging logic in the function is what we want users to remember. I would be interested to learn why you think it's too much? Maybe wording can be more concise?

Mental picture of experiment tracking in Ray Train

> ==================================================

Author: changed to flow based.

There are many experiment tracking frameworks out there. So instead of trying to 
cover every of them with a variety of usages, let’s first look at how to think 
about experiment tracking with Ray Train DataParallelTrainer, and how it may be 
different than running native training code in a single process. This mental 
picture helps you reason about some errors and makes debugging a lot easier.

In Ray Train DataParallelTrainer, interaction with experiment tracking services 
happens inside of ``train_loop_per_worker``. ``train_loop_per_worker`` is run as a separate process on 
an arbitrary node. This is different from single process training in one 
node in the following ways.

``train_loop_per_worker`` needs to have access to the tracking service
----------------------------------------------------------------------

> Usually this is trivial for online externally hosted tracking services. 

Reviewer: How many exp tracking tools are we going to discuss in this user guide? Wandb, MLFlow, Tensorboard? 

Do you think we need to ensure all sections have info for each framework? (We are missing wandb in this section.) And we can separate all by tabs?

Author: wandb is trivial for this section, which says a lot about its simplicity actually.
mlflow is difficult in the sense it has 4/5 flavors? It took me some time to set it up myself and I imagine the same for our users. Hence this section.

-------------------------------------------------------------

There are mainly two mechanisms when accessing credentials: 
checking some env var value, and checking the content of some config file. 
When applicable, you need to make sure that env var is set in the training 
function and the file is stored on shared storage to be accessible.

Avoid duplications
------------------

There are probably multiple ``train_loop_per_worker`` processes running, 
each corresponding to a Ray Train Worker. For distributed data 
parallel training, all of them should have the same results. 
You may want to only log from rank 0 worker to avoid duplications. 
You may find how to do it in the following code examples.

Code Example
============

In the following session, we will show some examples using Wandb 
and MLflow but the idea should be easily adaptable to other frameworks.

Pytorch
-------

Conceptual code snippets
~~~~~~~~~~~~~~~~~~~~~~~~

.. tabs::

    .. tab:: wandb

        .. code-block:: python
            
            from ray import train
            import wandb

            # assuming passing API key through config
            def train_func(config):

>                 os.environ["WANDB_API_KEY"] = config["wandb_api_key"]

Author: unclear. both are ok.

                wandb.init(
                    id=..., # or train.get_context().get_trial_id()
                    name=..., # or train.get_context().get_trial_name(),

>                     group=..., # or train.get_context().get_experiment_name(),

Reviewer 2: I would suggest moving the out, or making it clearer that these are suggested values.

Author: reworked the entire session above to be "steps" oriented instead of anchoring on dist vs non dist. Please take a look

                    # ...
                )
                # ...

                loss = optimize()

                metrics = {"loss": loss}
                # Only report the first worker results to wandb to avoid dup
                if train.get_context().get_world_rank() == 0:
                    wandb.log(metrics)

    .. tab:: file based MLflow

        .. code-block:: python
            
            from ray import train
            import mlflow

            # assuming passing a save dir through config
            def train_func(config):
                save_dir = config["save_dir"]
                mlflow.start_run(tracking_uri=f"file:{save_dir}")

                # ...

                loss = optimize()

                metrics = {"loss": loss}
                # Only report the first worker results to mlflow to avoid dup
                if train.get_context().get_world_rank() == 0:
                    mlflow.log_metrics(metrics)

>     .. tab:: MLflow externally hosted by databricks

Reviewer 2: Is it necessary to show this? Wondering how much value this provides, or if the user can just easily adapt from the previous example.

    the same run after restoration.

    For example:
    
    - `WandbLogger(id=UNIQUE_ID)`
    - `CometLogger(experiment_key=UNIQUE_ID)`
    - `MLFlowLogger(run_id=UNIQUE_ID)`

Legacy APIs
===========

Should I use ``setup_wandb`` and ``setup_mlflow``?
--------------------------------------------------

``setup_wandb`` and ``setup_mlflow`` are just convenient wrappers on top of
Weights&Biases and MLflow's native APIs. To spare you from learning yet another
set of APIs, we recommend you to continue using the native experiment tracking 
directly.
Just remember to guard with ``rank==0`` to avoid duplications whenever applicable.

Words on ``MLflowLoggerCallback`` and ``WandbLoggerCallback``
-------------------------------------------------------------

More seasoned Ray library users may wonder “Wait, what about `MLflowLoggerCallback
` and `WandbLoggerCallback`?” 
These APIs rely on reporting results to Ray Train and then log to experiment training 
frameworks. The usage scenario is not as versatile and requires some changes to the 
tracking logic. As a result, we advise users against using these callbacks when using any

> DataParallelTrainer.

Reviewer 2: Would personally prefer not to include these here, as it'll be more confusing for new users.

Author: Can we say something to guard against confusion? 
My worry about moving it entirely is the existing users. They are going to ask about these questions for sure.

Author: removed entirely

.. _train-experiment-tracking-native:

==================================

> Experiment Tracking with Ray Train

Reviewer: ```suggestion
Experiment Tracking
```
<img width="292" alt="Screenshot 2023-08-23 at 3 28 55 PM" src="https://github.com/ray-project/ray/assets/26745457/23e5a932-8551-426a-ab2a-563f51cf1e3d">

Is this more consistent with other titles?

Author: good point. updated!

specific training frameworks, for example ``mlflow.pytorch.autolog()``, 
``lightning.pytorch.loggers.MLFlowLogger`` etc. 
In this way, you get to port your experiment tracking logic to Ray Train with minimal change. 

How to set up your code to log distributed training experiment
==============================================================

Step 1: Decide a tracking service to use
----------------------------------------

For easy configuration, we recommend the following:

**MLflow:**

- offline mode: file based tracking uri pointing to shared storage path (``mlflow.start_run(tracking_uri=”file:some_shared_storage_path”)``)
- online mode: externally hosted by Databricks(``mlflow.start_run(tracking_uri=”databricks”)``).

**TensorBoard:**

A shared storage location where all nodes can write to.

**W&B:**

- offline mode: directory pointing to shared storage path (``wandb.init(dir="some_shared_storage_path")``)
- online mode: should work out-of-the-box with correct credentials

Step 2: Make sure that all nodes and processes have access to credentials
-------------------------------------------------------------------------

See example below.

> Step3: Initialize the run 

Reviewer: Can we add some simple code snippets for each step to make it clearer?

Author: I feel like I am repeating myself. The code snippet is shown down below. But I can touch on it a bit.

                # Only report the first worker results to wandb to avoid dup
                if train.get_context().get_world_rank() == 0:
                    wandb.log(metrics)

                # ...

                if train.get_context().get_world_rank() == 0:
                    wandb.finish()

    .. tab:: file based MLflow (offline)

        .. code-block:: python
            
            from ray import train
            import mlflow

            # assuming passing a save dir through config
            def train_func(config):
                save_dir = config["save_dir"]
                if train.get_context().get_world_rank() == 0:
                    # mlflow works the best if this is a folder dedicated to mlruns.
                    mlflow.set_tracking_uri(f"file:{save_dir}")
                    mlflow.set_experiment("my_experiment")
                    mlflow.start_run()

                # ...

                loss = optimize()

                metrics = {"loss": loss}
                # Only report the first worker results to mlflow to avoid dup
                if train.get_context().get_world_rank() == 0:
                    mlflow.log_metrics(metrics)

>     .. tab:: MLflow externally hosted by databricks (online)

Reviewer: The tab names are too long and wrapped to the next line, can we make it shorter(and switch the offline/online order)? like

```
Wandb(online), Wandb(offline), MLflow (online), MLflow(offline) 
```
<img width="781" alt="Screenshot 2023-08-23 at 3 37 59 PM" src="https://github.com/ray-project/ray/assets/26745457/096d4da4-3469-4541-b017-d3e3dba521db">

Author: Done!

Step 2: Make sure that all nodes and processes have access to credentials
-------------------------------------------------------------------------

**W&B:**

- :code:`wandb.login(key="your_api_key")` or :code:`os.environ["WANDB_API_KEY"] = "your_api_key"`

**MLflow:**

- This is done by having ``databrickscfg`` accessible by all nodes. See below for code snippet.

Step 3: Initialize the run 
--------------------------

Ray Train provides a training context that provides access to training identifiers. For example, 

* Training ID (:meth:`context.get_trial_id() <ray.train.context.TrainContext.get_trial_id>`) 
* Training Name (:meth:`context.get_trial_name() <ray.train.context.TrainContext.get_trial_name>`)

Step 4: Log
-----------

For some logs, every Worker generates an identical copy and saving a single copy is sufficient. 
Ray Train can save logs from the rank 0 Worker only with the following method:
:meth:`context.get_world_rank() <ray.train.context.TrainContext.get_world_rank>`.

Step 5: Finish the run
----------------------

For frameworks that require a call to mark a run as finished, include the appropriate call.
For example, ``wandb.finish()``.

> Code Example

Reviewer 3: ```suggestion
Code Examples
```

> ============

Reviewer 3: ```suggestion
=============
```

Let's see how the above works with some code.

The following session uses Wandb and MLflow but it is adaptable to other frameworks.

Pytorch
-------

Conceptual code snippets
~~~~~~~~~~~~~~~~~~~~~~~~

.. tabs::

>     .. tab:: Wandb(online)

Reviewer 3: What do you think about two layers of tabs so that we can keep "Weights & Biases" as a group-tab?

            :language: python
            :start-after: __lightning_experiment_tracking_comet_start__

    .. tab:: mlflow

        .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_mlflow.py
            :language: python
            :start-after: __lightning_experiment_tracking_mlflow_start__

    .. tab:: tensorboard
        
        .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_tensorboard.py
            :language: python
            :start-after: __lightning_experiment_tracking_tensorboard_start__
        
.. tip::
    
    When performing **fault-tolerant training** with auto-restoration, be sure 
    to specify a unique ID for the Loggers, so that the new workers report to
    the same run after restoration.

    For example:
    
    - `WandbLogger(id=UNIQUE_ID)`
    - `CometLogger(experiment_key=UNIQUE_ID)`
    - `MLFlowLogger(run_id=UNIQUE_ID)`

Common Errors
=============

**I have already called `wandb login` cli, but still getting 
"wandb: ERROR api_key not configured (no-tty). 
call wandb.login(key=[your_api_key])."**

> This is probably due to wandb credentials are not set up correctly

Reviewer 3: gentle ping on this suggestion!

on worker nodes. Make sure that you run ``wandb.login`` inside each
training function. You can take a look at the example above.

**"databricks_cli.utils.InvalidConfigurationError: 
You haven't configured the CLI yet!"**

> This is usually caused by running ``databricks configure`` which 

Reviewer 3: gentle ping on this suggestion!

                os.environ["WANDB_MODE"] = "offline"
                wandb.init(dir="some_shared_storage_path/wandb") 

    .. tab:: MLflow
        
        - online (hosted by Databricks)
            
            Ensure that all nodes have access to the Databricks config file.

            .. code-block:: python

                # MLflow client will look for a Databricks config file 
                # at the location specified by os.environ["DATABRICKS_CONFIG_FILE"].
                os.environ["DATABRICKS_CONFIG_FILE"] = config["databricks_config_file"]
                mlflow.set_tracking_uri("databricks")
                mlflow.start_run()

        - offline

            Ensure that offline directory points to a shared storage path.

            .. code-block:: python

                mlflow.start_run(tracking_uri="file:some_shared_storage_path/mlruns")

Setting up credentials
~~~~~~~~~~~~~~~~~~~~~~

Please refer to each tracking library's API documentation on this.
This usually involves setting some environment variable or accessing some config file.

For setting environment variable, one may pass the value of environment variable through the ``config`` 

> argument of ``train_func`` and set the corresponding environment variable in the ``train_func``.

Author: changed to use runtime_env

> For accessing config file, one needs to ensure that the config file is accessible to all nodes.

Author: One way to do this is by setting up a shared storage. Another way is to save a copy in each node. 
Added..

Setting up shared file system
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This involves setting up a network filesystem accessible to all nodes in the cluster, 
e.g. AWS EFS or Google Cloud Filestore.

Step 2: Configure and start the run 
-----------------------------------

This usually concerns picking an identifier for the run and associating it with a project.
Please refer to tracking libraries' own documentation for semantics. 

.. tip::
    
    When performing **fault-tolerant training** with auto-restoration, make sure that you use a 

>     consistent id to configure all tracking runs that logically belong to the same training run.

Reviewer 2: For this I think it's worthwhile to think about why we had the `setup` utilities, and convey to the user how to do these things themselves.

.. _train-experiment-tracking-native:

===================
Experiment Tracking
===================

.. note::

>     This guide is relevant for all DataParallelTrainer, including TorchTrainer and TensorflowTrainer.

Reviewer 4: ```suggestion
    This guide is relevant for all trainers in which a custom training loop is defined. This includes :class:`~ray.train.torch.TorchTrainer` and :class:`~ray.train.tensorflow.TensorflowTrainer`.
```

Most experiment tracking libraries work out-of-the-box with Ray Train. 
This guide provides instructions on how to set up the code so that your favorite experiment tracking libraries 
can work for distributed data parallel training with Ray Train. We will conclude the session with debugging

> tips.

Reviewer 4: Let's just call it "distributed training" - in model parallel training or a mix we also only want to report once

> Before we begin, the following is roughly how you can use the native experiment tracking lirary calls 

Reviewer 4: ```suggestion
Before we begin, the following is roughly how you can use the native experiment tracking library calls 
```

inside of Ray Train. 

.. code-block:: python

    from ray.train.torch import TorchTrainer
    from ray.train import ScalingConfig

    def train_func(config):
        # Training code and native lexperiment tracking library calls go here.

    scaling_config = ScalingConfig(num_workers=2, use_gpu=True)
    trainer = TorchTrainer(train_func, scaling_config=scaling_config)
    result = trainer.fit()

Ray Train lets you use native experiment tracking libraries by customizing the tracking 
logic inside the ``train_func`` function. In this way, you can port your experiment tracking 
logic to Ray Train with minimal changes. 

Log distributed training experiments
====================================

For some scenarios, every Worker generates an identical copy and saving a single copy is sufficient. 
Ray Train lets you apply logic to only the rank 0 Worker with the following method:

> :meth:`context.get_world_rank() <ray.train.context.TrainContext.get_world_rank>`.

Reviewer 4: I think we can work this out a bit more, as this is pretty central to the example code.

E.g. something like this
- In non-distributed training, you just use the regular tracking library API
- But in distributed training, many processes run the same code in parallel
- If all these processes report to the tracking library, you will end up with duplicated results
- So we usually only want to report from one of the processes.
- This is usually the "rank 0" process.

There are roughly two kinds of experiment tracking modes: online and offline. For online mode, you log
towards a tracking service that is running. Usually you need credentials to access the service.

> For offline mode, you log towards local file directory. Usually no credentials are needed.

Reviewer 4: I don't think this is very central to experiment tracking. I believe most users are using online logging, and we should optimize the guide for those.

It's still worth pointing out that all nodes should have access to the credentials

Author: I reworked this section quite a bit. it still touches upon offline a bit (just because I find some users do use this for development) but not too much.

For online mode, it is important to make sure that all nodes and Worker processes have 
access to credentials. For offline mode, major consideration is to ensure that there is a 
shared file system where all nodes can write to.

Step 1: Choose a tracking service to use

> ----------------------------------------

Reviewer 4: I think the five steps are too involved for the simple message that they convey.

I think we can simplify this with a short list:

1. Initialize the tracking service
2. Log results to the service
3. Clean up the session

and then use the short tabbed examples that we already have below to illustrate that

Author: ah yes, updated this section completely:

- Set up to connect to a tracking backend
- Configure and launch a run
- Log
- Finish the run

Most experiment tracking libraries work out-of-the-box with Ray Train. 
This guide provides instructions on how to set up the code so that your favorite experiment tracking libraries 
can work for distributed data parallel training with Ray Train. We will conclude the session with debugging
tips.

Before we begin, the following is roughly how you can use the native experiment tracking lirary calls 
inside of Ray Train. 

.. code-block:: python

    from ray.train.torch import TorchTrainer
    from ray.train import ScalingConfig

    def train_func(config):
        # Training code and native lexperiment tracking library calls go here.

    scaling_config = ScalingConfig(num_workers=2, use_gpu=True)
    trainer = TorchTrainer(train_func, scaling_config=scaling_config)
    result = trainer.fit()

Ray Train lets you use native experiment tracking libraries by customizing the tracking 
logic inside the ``train_func`` function. In this way, you can port your experiment tracking 
logic to Ray Train with minimal changes. 

Log distributed training experiments
====================================

> In this section, we break the interaction with experiment tracking backend into 4 logical steps:

Reviewer 2: As discussed, are you planning to move the conceptual code snippets up to this section?

- Set up to connect to a tracking backend
- Configure and launch a run
- Log
- Finish the run

Let's dive into each one of them.

.. note::

    For some scenarios, every Worker generates an identical copy and saving a single copy is sufficient. 
    Ray Train lets you apply logic to only the rank 0 worker with the following method:

>     :meth:`context.get_world_rank() <ray.train.context.TrainContext.get_world_rank>`.

Reviewer 2: This needs to be more explicit, I'm not sure what "copy" is referring to here. 

Maybe having the code snippet first will help make this clear what this is referring to.

Author: this is updated.

> Step 1: Set up necessary components to be able to connect to the tracking backend of your choice

Reviewer 2: This new section does a good job of encapsulating the offline/online (the user doesn't have to be aware about this at all in any other section anymore). With that, I really think we can further simplify this by not having to introduce these in the first place.

Concretely, I would recommend changing each tab to show a single code snippet example which outlines the different options (to make it more concrete to the user), and a section underneath that describes the concepts that are being used. For example - we can describe what shared storage is, or best practices for sharing keys with distributed workers.

            .. tab:: offline mode

                Make sure that :code:`os.environ["WANDB_MODE"] = "offline"` is set in ``train_func``.

                Set Wandb directory to point to a shared storage path: :code:`wandb.init(dir="some_shared_storage_path/wandb")` 

    .. tab:: MLflow

        .. tabs::

            .. tab:: online mode (hosted by Databricks)
                
                Start the run with :code:`mlflow.start_run(tracking_uri="databricks")`

                Make sure that all nodes have access to ``databrickscfg`` file.

            .. tab:: offline mode

                Start the run by setting tracking uri to a shared storage path: 
                :code:`mlflow.start_run(tracking_uri="file:some_shared_storage_path/mlruns")`

    .. tab:: TensorBoard (offline)
        
        Set up ``SummaryWriter`` to write to a shared storage path: :code:`writer = SummaryWriter("some_shared_storage_path/runs")`

Step 2: Initialize the run 
--------------------------

Ray Train provides a training context that provides access to training identifiers. For example, 

* Training ID (:meth:`context.get_trial_id() <ray.train.context.TrainContext.get_trial_id>`) 

> * Training Name (:meth:`context.get_trial_name() <ray.train.context.TrainContext.get_trial_name>`)

Reviewer 2: It would be helpful to describe more in detail about the usage of these - are these unique identifiers of the run, an experiment, or something else?

.. tip::
    
    When performing **fault-tolerant training** with auto-restoration, be sure 

>     to specify a unique ID for the Loggers, so that the new workers report to

Reviewer 2: nit: I know we shuffled this so it used to be pertaining to Lightning, but I think it would be preferable to not use Logger here, and only introduce that in the Lightning section below.

    the same run after restoration.

    For example:
    
    - `WandbLogger(id=UNIQUE_ID)`
    - `CometLogger(experiment_key=UNIQUE_ID)`

>     - `MLFlowLogger(run_id=UNIQUE_ID)`

Reviewer 2: Similarly, are these referencing the Training ID or something else? As a user how do I tie this with the above?

Can I use the trial ID here?

Author: The `UNIQUE_ID` doesn't have to be Training ID, it can be anything that's unique tbh...
I sense this is confusing to people. Maybe we should remove the suggestions on what can go in as `UNIQUE_ID`?
It really can simply just `uuid.uuid4()`.

Step 3: Log
-----------

You can customize when and where to log parameters, metrics, models, or media contents. 
You can also use some native integrations that these tracking frameworks have with 
specific training frameworks, for example ``mlflow.pytorch.autolog()``, 
``lightning.pytorch.loggers.MLFlowLogger`` etc. 

Step 4: Finish the run
----------------------

For frameworks that require a call to mark a run as finished, include the appropriate call.

> For example, ``wandb.finish()``.

Reviewer 2: This is a little terse right now and I'd like to make sure it's clear to the user why this information is here. For example, we have finishing the run as a step here because it's needed to ensure artifacts are uploaded (or something like this, right?)

Code snippets might help.

Conceptual code snippets
========================

Let's see how the above works with some code.

The following session uses Wandb and MLflow but it is adaptable to other frameworks.

.. tabs::

    .. tab:: Wandb(online)

        .. code-block:: python
            
            from ray import train
            import wandb

            # Assumes you are passing API key through config
            def train_func(config):
                if train.get_context().get_world_rank() == 0:
                    wandb.login(key=config["wandb_api_key"])

                    wandb.init(
                        id=..., # or train.get_context().get_trial_id(),

>                         name=..., # or train.get_context().get_trial_name(),

Reviewer 2: I don't think "or" makes sense here, maybe "Example: " instead

>                         group=..., # or train.get_context().get_experiment_name(),

Reviewer 2: nit: Do we want people to use this? Or only if they're using Tune?

In this section, we break the interaction with experiment tracking backend into 4 logical steps:
- Set up to connect to a tracking backend
- Configure and launch a run
- Log
- Finish the run

Let's dive into each one of them.

.. note::

    A major difference between distributed and non-distributed training is that in distributed training, 
    multiple processes are running in parallel and under certain setups they have the same results. If all 
    of them are reported to the tracking backend, there will be duplicated results. To address that,  
    Ray Train lets you apply logging logic to only the rank 0 worker with the following method:
    :meth:`context.get_world_rank() <ray.train.context.TrainContext.get_world_rank>`.

    .. code-block:: python

        from ray import train
        def train_func(config):
            ...
            if train.get_context().get_world_rank() == 0:
                # do your logging logic only for rank0 worker.
            ...

Step 1: Set up necessary components to be able to connect to the tracking backend of your choice
------------------------------------------------------------------------------------------------

> First, you should choose which tracking backend to use: W&B, MLflow, TensorBoard etc.

Reviewer 2: For this section would it be helpful to showcase all the tracking services you use in this document? Specifically, Comet is referenced later but is the only one not mentioned in this introduction section.

    .. code-block:: python

        from ray import train
        def train_func(config):
            ...
            if train.get_context().get_world_rank() == 0:
                # do your logging logic only for rank0 worker.
            ...

The interaction with experiment tracking backend within the ``train_func`` can be broken 
into 4 logical steps:

- Set up to connect to a tracking backend
- Configure and launch a run
- Log
- Finish the run

Let's dive into each one of them.

Step 1: Set up necessary components to connect to the tracking backend of your choice
-------------------------------------------------------------------------------------

First, you should choose which tracking backend to use: W&B, MLflow, TensorBoard etc.

Some of them operate under either online or offline mode, each with different considerations when
being set up.
For online mode, you log towards a tracking service that is running. Usually you need credentials to access the service.
Under this mode, you need to ensure that all nodes and worker processes have access to credentials.
For offline mode, you log towards the local file directory. Usually no credentials are needed. You need to instead

> ensure that there is a shared file system where all nodes can write to.

Reviewer 2: This can be condensed to just say you need to make sure you are authenticated properly on the distributed worker

.. tabs::

    .. tab:: Wandb

        **online**

        Ensure that credentials are set inside of ``train_func``.

        .. code-block:: python
            
            # This is equivalent to `os.environ["WANDB_API_KEY"] = "your_api_key"`
            wandb.login(key="your_api_key")

        **offline**

        Ensure that offline directory points to a shared storage path.

        .. code-block:: python

            os.environ["WANDB_MODE"] = "offline"

>             wandb.init(dir="some_shared_storage_path/wandb") 

Reviewer 2: Can you define Online and Offline directly in here? For online, this means using a W&B API Key to authenticate to the serve, and log to W&B server. For offline, this means no authentication and logging directly to shared file system as the backend.

This would
1: Make it clear how this maps to the next subsections (setting up credentials, setting up shared storage)
2. Allow you to use MLflow terminology (local, remote) rather than online/offline.

    .. tab:: MLflow
        
        **online (hosted by Databricks)**
            
        Ensure that all nodes have access to the Databricks config file.

        .. code-block:: python

            # MLflow client will look for a Databricks config file 
            # at the location specified by `os.environ["DATABRICKS_CONFIG_FILE"]`.
            os.environ["DATABRICKS_CONFIG_FILE"] = config["databricks_config_file"]
            mlflow.set_tracking_uri("databricks")
            mlflow.start_run()

        **offline**

        Ensure that offline directory points to a shared storage path.

        .. code-block:: python

            mlflow.start_run(tracking_uri="file:some_shared_storage_path/mlruns")

Setting up credentials
~~~~~~~~~~~~~~~~~~~~~~

Please refer to each tracking library's API documentation on this.
This usually involves setting some environment variable or accessing some config file.

For setting environment variable, one may pass the value of environment variable through the ``config`` 

> argument of ``train_func`` and set the corresponding environment variable in the ``train_func``.

Reviewer 2: Update this section to explain how to use environment variables.

For accessing config file, one needs to ensure that the config file is accessible to all nodes.
One way to do this is by setting up a shared storage. Another way is to save a copy in each node.

Setting up shared file system
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This involves setting up a network filesystem accessible to all nodes in the cluster, 
e.g. AWS EFS or Google Cloud Filestore.

Step 2: Configure and start the run 
-----------------------------------

> This usually concerns picking an identifier for the run and associating it with a project.

Reviewer 2: One thing we discussed yesterday was about how to link this experiment with the Ray Train results folder/metadata. You mentioned that you didn't want to use the trial name, but would prefer to recommend logging the metadata as part of the config instead. Can you add that?

Please refer to tracking libraries' own documentation for semantics. 

.. tip::
    
    When performing **fault-tolerant training** with auto-restoration, make sure that you use a 
    consistent id to configure all tracking runs that logically belong to the same training run.
    One way to acquire an unique id is through 

>     :meth:`ray.train.get_context().get_trial_id() <ray.train.TrainContext.get_trial_id>`.

Reviewer 2: This isn't linking properly for some reason.

    .. code-block:: python

        import ray
        from ray.train import ScalingConfig, RunConfig, FailureConfig
        from ray.train.torch import TorchTrainer

        def train_func(config):
            if ray.train.get_context().get_world_rank() == 0:
                wandb.init(id=ray.train.get_context().get_trial_id())
            ...

        trainer = TorchTrainer(
            train_func, 
            run_config=RunConfig(failure_config=FailureConfig(max_failures=3))
        )

        trainer.fit()
            

Step 3: Log
-----------

You can customize within ``train_func`` how to log parameters, metrics, models, or media contents 
just as you would with a single process training script. 
You can also use native integrations that a particular tracking framework has with 
specific training frameworks, for example ``mlflow.pytorch.autolog()``, 
``lightning.pytorch.loggers.MLFlowLogger`` etc. 

Step 4: Finish the run
----------------------

This step ensures that all logs are synced to the tracking service. Depending on the implementation of 
various tracking libraries, sometimes logs are first stored locally before they are synced to the tracking 

> service. When training is finished, the node that the training worker resided on could be turned down by 
> ray autoscaler. Finishing the run makes sure that all logs are synced by the time training workers exit. 

Reviewer 2: ```suggestion
service. Finishing the run makes sure that all logs are synced by the time training workers exit. 
```

There's no need to mention the node/autoscaler. I think more technically this happens when the Ray Actor terminates, but the important part for the user to know is that this will sync all the data before the training finishes.

**Wandb**

.. code-block:: python

    # https://docs.wandb.ai/ref/python/finish
    wandb.finish()

**MLflow**

.. code-block:: python

    # https://mlflow.org/docs/1.2.0/python_api/mlflow.html
    mlflow.end_run()

**Comet**

.. code-block:: python

    # https://www.comet.com/docs/v2/api-and-sdk/python-sdk/reference/Experiment/#experimentend
    Experiment.end()

> Runnable Code
> =============

Reviewer 2: ```suggestion
Examples
========
```

PyTorch
-------

.. dropdown:: Log to Wandb (online) 

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking//torch_exp_tracking_wandb.py
            :emphasize-lines: 13, 14, 18, 19, 48, 49, 51, 52
            :language: python
            :start-after: __start__

.. dropdown:: Log to file based MLflow (offline)         

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/torch_exp_tracking_mlflow.py
        :emphasize-lines: 22, 23, 54, 55, 57, 58, 64
        :language: python
        :start-after: __start__
        :end-before: __end__

PyTorch Lightning
-----------------

The native Logger integration in PyTorch Lightning with W&B, CometML, MLFlow, 
and Tensorboard can still be used seamlessly with Ray Train TorchTrainer.

The following example will walk you through how. The code here is runnable. 
There is a common shared piece of setting up a dummy model and dataloader
just for demonstration purposes.
        

> .. dropdown:: Define your model and dataloader (Dummy ones for demonestration purposes)

Reviewer 2: Should we just include this in each collapsible block instead? I think as a user it's easier to just click open one block.

        MLflow offers both *local* and *remote* modes. For *local* mode, you log towards a local file 
        system. In distributed training, this should be a shared storage path that all nodes can write 
        to (More information `here<setting-up-shared-file-system>`). 
        For *remote* mode, people usually choose to log to Databrick's MLflow service. Proper credentials 
        need to be set for each training worker (More information `here<setting-up-credentials>`).

        **offline**

        Ensure that offline directory points to a shared storage path.

        .. code-block:: python

            mlflow.start_run(tracking_uri="file:some_shared_storage_path/mlruns")

        **online (hosted by Databricks)**
            
        Ensure that all nodes have access to the Databricks config file.

        .. code-block:: python

            # MLflow client will look for a Databricks config file 
            # at the location specified by `os.environ["DATABRICKS_CONFIG_FILE"]`.
            os.environ["DATABRICKS_CONFIG_FILE"] = config["databricks_config_file"]
            mlflow.set_tracking_uri("databricks")
            mlflow.start_run()

.. setting-up-credentials

> Setting up credentials

Reviewer 3: ```suggestion
Set up credentials
```

    wandb.finish()

**MLflow**

.. code-block:: python

    # https://mlflow.org/docs/1.2.0/python_api/mlflow.html
    mlflow.end_run()

**Comet**

.. code-block:: python

    # https://www.comet.com/docs/v2/api-and-sdk/python-sdk/reference/Experiment/#experimentend
    Experiment.end()

Examples
========

PyTorch
-------

.. dropdown:: Log to Wandb

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking//torch_exp_tracking_wandb.py
            :emphasize-lines: 13, 14, 18, 19, 48, 49, 51, 52
            :language: python
            :start-after: __start__

.. dropdown:: Log to file based MLflow

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/torch_exp_tracking_mlflow.py
        :emphasize-lines: 22, 23, 54, 55, 57, 58, 64
        :language: python
        :start-after: __start__
        :end-before: __end__

PyTorch Lightning
-----------------

The native Logger integration in PyTorch Lightning with **W&B**, **CometML**, **MLFlow**, 
and **Tensorboard** can still be used seamlessly with Ray Train TorchTrainer.

The following example will walk you through how. The code here is runnable. 

> There is a common shared piece of setting up a dummy model and dataloader

Reviewer 3: Use this code to set up a dummy model and dataloader...?

Reviewer 3: This sentence is a bit awkward. We can discuss IRL.

        :language: python

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_comet.py
            :language: python
            :start-after: __lightning_experiment_tracking_comet_start__

.. dropdown:: TensorBoard
  
    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_model_dl.py
        :language: python

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_tensorboard.py
            :language: python
            :start-after: __lightning_experiment_tracking_tensorboard_start__
            :end-before: __lightning_experiment_tracking_tensorboard_end__

Common Errors
=============

Missing Credentials
-------------------

**I have already called `wandb login` cli, but still getting 
"wandb: ERROR api_key not configured (no-tty). 
call wandb.login(key=[your_api_key])."**

This is probably due to wandb credentials are not set up correctly
on worker nodes. Make sure that you run ``wandb.login`` 
or pass ``WANDB_API_KEY`` to each training function. 

> See `"setting up credentials" section <setting-up-credentials>` for more details.

Reviewer 3: ```suggestion
See `Set up credentials <setting-up-credentials>` for more details.
```

Missing Configurations
----------------------

**"databricks_cli.utils.InvalidConfigurationError: 
You haven't configured the CLI yet!"**

This is usually caused by running ``databricks configure`` which 
generates ``~/.databrickscfg`` only on head node. Move this file to a shared
location or copy it to each node.

> See `"setting up credentials" section <setting-up-credentials>` for more details.

Reviewer 3: If you accept my suggestion to rename the section to "set up credentials", then this needs to be updated.

> See `"setting up credentials" section <setting-up-credentials>` for more details.

Reviewer 3: ```suggestion
See `Set up credentials <setting-up-credentials>` for more details.
```

the setup.

The following psudo code demonstrates how to use the native experiment tracking library calls 
inside of Ray Train: 

.. code-block:: python

    from ray.train.torch import TorchTrainer
    from ray.train import ScalingConfig

    def train_func(config):
        # Training code and native experiment tracking library calls go here.

    scaling_config = ScalingConfig(num_workers=2, use_gpu=True)
    trainer = TorchTrainer(train_func, scaling_config=scaling_config)
    result = trainer.fit()

Ray Train lets you use native experiment tracking libraries by customizing the tracking 
logic inside the ``train_func`` function. In this way, you can port your experiment tracking 
logic to Ray Train with minimal changes. 

Getting Started
===============

Let's start by looking at some code snippets.

The following session uses Weights & Biases (W&B) and MLflow but it is adaptable to other frameworks.

.. tabs::

    .. tab:: W&B

        .. code-block:: python
            
            import ray
            from ray import train
            import wandb

            # This ensures that all ray worker processes have `WANDB_API_KEY` set.

>             ray.init(runtime_env={"env_vars": {"WANDB_API_KEY": "your_api_key"}})

Reviewer 2: step 1 here

            def train_func(config):
                # Step 1 and 2
                if train.get_context().get_world_rank() == 0:
                    wandb.init(
                        name=...,
                        project=...,

>                         # ...

Reviewer 2: Use `id` here since it's what you refer to in step 2.

Author: I don't think it's a big deal. name and project etc are just more intuitive when referencing the run. For that reason, that's what you see if you open wandb website and click only one example. And we should try to keep that intuitiveness. 
id is what you do when you care about logically continue the last run (a tip about fault tolerant, a more advanced case).

            # mv ~/.databrickscfg YOUR_SHARED_STORAGE_PATH
            # This function assumes `databricks_config_file` in config
            def train_func(config):
                # Step 1 and 2
                os.environ["DATABRICKS_CONFIG_FILE"] = config["databricks_config_file"]
                mlflow.set_tracking_uri("databricks")
                mlflow.set_experiment_id(...)
                mlflow.start_run()

                # ...

                loss = optimize()

                metrics = {"loss": loss}
                # Only report the results from the first worker to MLflow to avoid duplication

                # Step 3
                if train.get_context().get_world_rank() == 0:
                    mlflow.log_metrics(metrics)

.. tip::

    A major difference between distributed and non-distributed training is that in distributed training, 
    multiple processes are running in parallel and under certain setups they have the same results. If all 
    of them report results to the tracking backend, you may get duplicated results. To address that,  
    Ray Train lets you apply logging logic to only the rank 0 worker with the following method:

>     :meth:`context.get_world_rank() <ray.train.context.TrainContext.get_world_rank>`.

Reviewer 2: ```suggestion
    :meth:`ray.train.get_context().get_world_rank() <ray.train.context.TrainContext.get_world_rank>`.
```

        from ray import train
        def train_func(config):
            ...
            if train.get_context().get_world_rank() == 0:
                # do your logging logic only for rank0 worker.
            ...

The interaction with the experiment tracking backend within the ``train_func`` has 4 logical steps:

#. Set up the connection to a tracking backend
#. Configure and launch a run
#. Log metrics
#. Finish the run

More details about each step follows.

Step 1: Connect to your tracking backend
----------------------------------------

First, decide which tracking backend to use: W&B, MLflow, TensorBoard, Comet, etc.
If applicable, make sure that you properly set up credentials on each training worker.

.. tabs::

    .. tab:: W&B
        
        W&B offers both *online* and *offline* modes. For *online* mode, because you log to W&B's
        tracking service, ensure that you set up correct credentials on each training worker 
        (See :ref:`Set up credentials<set-up-credentials>` for more information). 
        For *offline* mode, because you log towards a local file system, 
        make sure that the file system is a shared storage path that all nodes can write to 

>         (See :ref:`Set up a shared file system<set-up-shared-file-system>` for more information).

Reviewer 2: Move these directly into `online` and `offline`?

Reviewer 2: Same for MLflow

>         **online**

Reviewer 2: nit:
```suggestion
        **Online**
```

Reviewer 2: Same for others

..                 wandb.init(..., config={"ray_train_persistent_storage_path": "TODO: fill in when API stablizes"})

.. tip::
    
    When performing **fault-tolerant training** with auto-restoration, use a 
    consistent ID to configure all tracking runs that logically belong to the same training run.
    One way to acquire an unique ID is with the following method: 
    :meth:`ray.train.get_context().get_trial_id() <ray.train.context.TrainContext.get_trial_id>`.

    .. code-block:: python

        import ray
        from ray.train import ScalingConfig, RunConfig, FailureConfig
        from ray.train.torch import TorchTrainer

        def train_func(config):
            if ray.train.get_context().get_world_rank() == 0:
                wandb.init(id=ray.train.get_context().get_trial_id())
            ...

        trainer = TorchTrainer(
            train_func, 
            run_config=RunConfig(failure_config=FailureConfig(max_failures=3))
        )

        trainer.fit()
            

Step 3: Log metrics
-------------------

You can customize how to log parameters, metrics, models, or media contents, within ``train_func``,  

> just as in a single process training script. 

Reviewer 2: ```suggestion
just as in a non-distributed training script. 
```

You can also use native integrations that a particular tracking framework has with 
specific training frameworks. For example ``mlflow.pytorch.autolog()``, 
``lightning.pytorch.loggers.MLFlowLogger``, etc. 

Step 4: Finish the run
----------------------

This step ensures that all logs are synced to the tracking service. Depending on the implementation of 
various tracking libraries, sometimes logs are first cached locally and only synced to the tracking 
service in an asynchronous fashion. 
Finishing the run makes sure that all logs are synced by the time training workers exit. 

> **W&B**

Reviewer 2: nit: Use tabs

    # https://www.comet.com/docs/v2/api-and-sdk/python-sdk/reference/Experiment/#experimentend
    Experiment.end()

Examples
========

The following are runnable examples for PyTorch and PyTorch Lightning.

PyTorch
-------

.. dropdown:: Log to W&B

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking//torch_exp_tracking_wandb.py
            :emphasize-lines: 13, 14, 18, 19, 48, 49, 51, 52
            :language: python
            :start-after: __start__

.. dropdown:: Log to file based MLflow

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/torch_exp_tracking_mlflow.py
        :emphasize-lines: 22, 23, 24, 25, 54, 55, 57, 58, 64
        :language: python
        :start-after: __start__
        :end-before: __end__

PyTorch Lightning
-----------------

You can use the native Logger integration in PyTorch Lightning with **W&B**, **CometML**, **MLFlow**, 
and **Tensorboard**, while using Ray Train's TorchTrainer.

The following example walks you through the process. The code here is runnable. 
There is a common shared piece of setting up a dummy model and dataloader
just for demonstration purposes.

> .. dropdown:: Wandb

Reviewer 2: ```suggestion
.. dropdown:: W&B
```

            :language: python
            :start-after: __lightning_experiment_tracking_mlflow_start__
            :end-before: __lightning_experiment_tracking_mlflow_end__

.. dropdown:: Comet

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_model_dl.py
        :language: python

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_comet.py
            :language: python
            :start-after: __lightning_experiment_tracking_comet_start__

.. dropdown:: TensorBoard
  
    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_model_dl.py
        :language: python

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_tensorboard.py
            :language: python
            :start-after: __lightning_experiment_tracking_tensorboard_start__
            :end-before: __lightning_experiment_tracking_tensorboard_end__

Common Errors
=============

Missing Credentials
-------------------

> **I have already called `wandb login` cli, but still getting 

Reviewer 2: nit: The formatting of this message and the one in the section below are slightly different. I would recommend making them consistent and putting the actual error message in a code block for readability purposes.

            :language: python
            :start-after: __lightning_experiment_tracking_mlflow_start__
            :end-before: __lightning_experiment_tracking_mlflow_end__

.. dropdown:: Comet

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_model_dl.py
        :language: python

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_comet.py
            :language: python
            :start-after: __lightning_experiment_tracking_comet_start__

.. dropdown:: TensorBoard
  
    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_model_dl.py
        :language: python

    .. literalinclude:: ../../../../python/ray/train/examples/experiment_tracking/lightning_exp_tracking_tensorboard.py
            :language: python
            :start-after: __lightning_experiment_tracking_tensorboard_start__
            :end-before: __lightning_experiment_tracking_tensorboard_end__

Common Errors
=============

Missing Credentials
-------------------

> **I have already called `wandb login` cli, but still getting** 

Reviewer 3: ```suggestion
**I have already called `wandb login` cli, but am still getting** 
```

.. code-block:: none

    wandb: ERROR api_key not configured (no-tty). call wandb.login(key=[your_api_key]).

This is probably due to wandb credentials are not set up correctly
on worker nodes. Make sure that you run ``wandb.login`` 
or pass ``WANDB_API_KEY`` to each training function. 
See :ref:`Set up credentials <set-up-credentials>` for more details.

Missing Configurations
----------------------

> **I have already run `databricks configure`, but still getting**

Reviewer 3: ```suggestion
**I have already run `databricks configure`, but am still getting**
```
